lms/djangoapps/teams/csv.py

```diff
@@ -243,16 +261,18 @@ def validate_header(self, header):
             return False
         return True
 
-    def validate_teamsets(self):
+    def validate_teamsets(self, csv_data):
         """
         Validates team set ids. Returns true if there are no errors.
         The following conditions result in errors:
         Teamset does not exist
         Teamset id is duplicated
         """
+        teamset_ids = csv_data.fieldnames[2:]
         valid_teamset_ids = {ts.teamset_id for ts in self.course.teams_configuration.teamsets}
+
         dupe_set = set()
-        for teamset_id in self.teamset_ids:
```

Comment:
I can see why you did it, by we already have `self.teamset_ids`. So you are passing `csv_data` just to look uniform?

Reply:
Yeah, I wanted to make the whole process more function-based. We have this mess right now where everything must be run in a specific order to build up the state and class variables to operate on and it's not at all testable. Moving to this paradigm, we can operate only on data that is passed in instead of relying on something to set the `self.teamset_ids` which vastly improves modularity and testing.

Reply:
I would disagree about `specific order` :) The state method are run at the top to initialize the variables.

Reply:
Still, it means I can't test any sub-function w/in `set_team_memberships` without manually invoking those functions or setting up the context. This gives us much better modularity which should improve testability.